app/src/main/java/org/worldbank/transport/driver/services/DriverLocationService.java

```diff
@@ -0,0 +1,465 @@
+            return DriverLocationService.this;
+        }
+    }
+
+    private LocationManager locationManager;
+    private LocationListener locationListener;
+    private Context context;
+    private final IBinder binder = new LocationServiceBinder();
+    private DriverLocationUpdateListener driverLocationUpdateListener;
+
+    // will be non-null when at least one location update received with the minimum accuracy
+    private Location acceptableLocation;
+    private long startedWaitingForBetterLocation;
+    private ArrayList<Location> preferredAccuracyLocations;
+    private boolean doneWaitingForLocations;
+
+    @Override
+    public void onConnected(@Nullable Bundle bundle) {
+        // Android Studio will complain mightily if requestLocationUpdates is not called
+        // after a permissions check that is in the same method
+        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
+            Log.d(LOG_LABEL, "Actually starting location updates now.");
+            locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 0, 0, locationListener);
+        } else {
+            Log.w(LOG_LABEL, "Got onConnected callback, but do not have location services permissions now");
+        }
+    }
+
+    @Override
+    public void onConnectionSuspended(int i) {
+        Log.w(LOG_LABEL, "Location services API connection suspended!");
+        // TODO: anything?
+    }
+
+    @Override
+    public void onConnectionFailed(ConnectionResult connectionResult) {
+        Log.w(LOG_LABEL, "Location services API connection failed!");
+        // TODO: anything?
+    }
+
+    /**
+     * Return best location estimate found for received updates.
+     * Should stop this service after receiving this result (can be accomplished by unbinding)
+     *
+     * @return best estimate of current location during service run
+     */
+    @Nullable
+    public Location getBestLocationFound() {
+
+        if (preferredAccuracyLocations.size() > 0) {
+            // calculate weighted average location from those found
+
+            if (preferredAccuracyLocations.size() == 1) {
+                // if only got one, just use it
+                return new Location(preferredAccuracyLocations.get(0));
+            }
+
+            Log.d(LOG_LABEL, "Finding weighted average of " + preferredAccuracyLocations.size() + " location readings");
+
+            double sumLat = 0;
+            double sumLon = 0;
+            float sumAccuracy = 0;
+            double sumInvertedAccuracy = 0;
+
+            for (Location location : preferredAccuracyLocations) {
+                float accuracy = location.getAccuracy();
+
+                // accuracy for these readings is within range of 0 to PREFERRED_LOCATION_ACCURACY
+                // smaller is better; invert range for weighted average
+                float invertedAccuracy = PREFERRED_LOCATION_ACCURACY - accuracy;
```

Comment:
The accuracy represents [the radius of 68% confidence](http://developer.android.com/reference/android/location/Location.html#getAccuracy%28%29). So about a third of the time, the location actually falls outside of the stated radius (the stated accuracy is itself inaccurate). As time passes, the GPS settles and the accuracy radius readings fall.